### ISSUE TITLE
Proposal for hook into resolution attempts of already resolved/rejected promise

### ISSUE THREAD

**Issue by bmeurer** [nodejs/promises-debugging#8]
As discussed during the summit, V8 could provide a hook to notify the embedder (Node/Chrome) about swallowed promise rejections, specifically with the `Promise` constructor (see [Use Case #3](https://github.com/nodejs/promise-use-cases/blob/master/use-cases/warnings/3/3.md) in [promise-use-cases](https://github.com/nodejs/promise-use-cases)). I've written a design document how to extend the V8 C++ API [here](https://docs.google.com/document/d/1ZDafLmQmvP4LgQNtUyLWaSgyhXwgr5nJ624eON0ix5w).

cc @MayaLekova @benjamingr

**Comment by benjamingr** [comment]
The hook sounds good to me (Confirming Node.js can do it and that it is sufficient to address the use case in question and expose a hook from Node's side).

Regarding the API - I'm not sure it should be the same function since no actual promise rejection is happening .

**Comment by bmeurer** [comment]
It'd be convenient to re-use the existing C++ API, since it exposes everything we need. But that's not a requirement, we can also introduce a new C++ API if you think that's better.

**Comment by benjamingr** [comment]
@bmeurer I think it's confusing to reuse the hook since at the moment it's only called when a promise rejects but I don't have very strong feelings about it.

Also cc @misterdjules who has raised this concern about promises before and @getify who raised this issue as well.

**Comment by bmeurer** [comment]
@benjamingr sounds reasonable. I added another possible [solution](https://docs.google.com/document/d/1ZDafLmQmvP4LgQNtUyLWaSgyhXwgr5nJ624eON0ix5w/edit#heading=h.gwn299idykxd) which would introduce a dedicated C++ hook just for this case.

**Comment by misterdjules** [comment]
Without the ability to differentiate between programmer errors (bugs that materialize as an uncaught exception) and operational errors (`reject` is called with an error object), I'm not sure it's going to be possible for the embedder or a developer to determine whether the process should crash or not.

Would it be possible to add the ability to differentiate between those two different use cases?

**Comment by benjamingr** [comment]
> Would it be possible to add the ability to differentiate between those two different use cases?

Unfortunately programmer errors and operational errors in the Promise constructor are not always distinguishable by whether or not it's a `throw` or a `reject`. I've seen multiple projects `throw` in the `Promise` constructor for operational stuff explicitly when doing the research for the summit.

Note that this hook is specifically for multiple `throw`s/`reject`s and the current hook for promise rejections doesn't create a distinction. The line gets even blurrier in `async` functions (although that's not this issue). 

I would love discussing how we can better distinguish programmer/operational errors and be more helpful towards programmers. (Although I think that should be a separate issue IMO).

If you think there is value in the promise rejection hook exposing whether or not a promise was `throw`n and not `reject`ed for APMs or debugging then that's also worth discussing (again, I believe in a separate issue).

----

If I understand it correctly - this issue is _specifically_ about detecting multiple `reject`s or `throw`s in the promise constructor and providing a hook for that that APMs can hook into or Node.js can warn on.

**Comment by misterdjules** [comment]
> Unfortunately programmer errors and operational errors in the Promise constructor are not always distinguishable by whether or not it's a throw or a reject. I've seen multiple projects throw in the Promise constructor for operational stuff explicitly when doing the research for the summit.

That's good to know. Do you have more details to share about that specific part of your research (how it was done, your findings, etc.)? I think it could be useful to include in one of the documents in this repository for future reference.

> Note that this hook is specifically for multiple throws/rejects and the current hook for promise rejections doesn't create a distinction. The line gets even blurrier in async functions (although that's not this issue).

Ah you're right, I was confused by the title of the issue. Reading the design document it's now clear that this is about a very specific use case. @bmeurer Do you think renaming this issue as "Proposal for hook into already resolved/rejected promise" would help clarify this?

I agree in this case it's clear that there is a bug in the application and that developers/embedders can take appropriate action, including crashing.

I also agree that being able to differentiate between promises rejected due to programmer or operational error should be discussed in a different issue.

Thanks for the clarification @benjamingr!

**Comment by bmeurer** [comment]
@misterdjules Added "Proposal for hook into already resolved/rejected promise" as subtitle to the design document. 👍

**Comment by getify** [comment]
> this issue is specifically about detecting multiple rejects or throws in the promise constructor

Is there a separate issue for catching exceptions that happened in the promise constructor without the use of `throw`, like calling `doesNotExist()`? That is absolutely an important use-case for me.

> Unfortunately programmer errors and operational errors in the Promise constructor are not always distinguishable by whether or not it's a throw or a reject.

Certainly that line may be blurry. But the line that shouldn't be blurry is whether an exception came from an explicit `throw` or whether it came from a runtime mistake in the programming. THAT distinction is very important to me. If that's a different issue or hook, fine. But I just wanted to make sure that wasn't getting glossed over.

**Comment by mhdawson** [comment]
Is it also possible to resolve after rejecting, needing a 

 kPromiseResolvedAfterReject = 3

**Comment by bmeurer** [comment]
@mhdawson I guess that'd be `kPromiseResolveAfterResolved`. The terminology is a bit odd here, but *resolved* for a promise actually means that it's no longer *pending*.

**Comment by bmeurer** [comment]
So as a promise state *resolved* means either *fulfilled* or *rejected*.

**Comment by benjamingr** [comment]
@bmeurer actually a promise _resolved_ means either fulfilled, rejected or tracking another promise. All three are resolved:

```js
var p = Promise.resolve(); // resolved and fulfilled
var p2 = Promise.reject(new Error()); // resolved and rejected
var p3 = Promise.resolve(new Promise(() => {})); // p3 is resolved
```

Resolved basically means "resolving it again would have no effect" in the context of this conversation.

Fullfilled or rejected is called _settled_.

**Comment by bmeurer** [comment]
@benjamingr Right. So the actual state of the promise can be queried from the object passed to the C++ API hook in case Node needs to know. Having `kPromiseResolveAfterResolved` and `kPromiseRejectAfterResolved` should be sufficient then on our side?

**Comment by benjamingr** [comment]
Correct :)

**Comment by bmeurer** [comment]
The [Swallowed Rejection Hook](https://docs.google.com/document/d/1ZDafLmQmvP4LgQNtUyLWaSgyhXwgr5nJ624eON0ix5w) just landed in V8 today ([CL](https://chromium-review.googlesource.com/c/v8/v8/+/1126099), [v8:7919](http://crbug.com/v8/7919)). We went for [Solution 1](https://docs.google.com/document/d/1ZDafLmQmvP4LgQNtUyLWaSgyhXwgr5nJ624eON0ix5w/edit#heading=h.7l7h78ogejt9) to keep the C++ API surface small.

Not sure how urgent this is, but the changes are relatively easy, so you could even consider floating this as patch on top of Node 10 if you want the warning there.

**Comment by getify** [comment]
> Is there a separate issue for catching exceptions that happened in the promise constructor without the use of throw, like calling doesNotExist()? That is absolutely an important use-case for me.

Ping.

**Comment by bmeurer** [comment]
@getify To clarify: You're talking about exceptions that are raised inside of the executor function, but not via a syntactic throw inside the executor?

**Comment by getify** [comment]
@bmeurer Correct... IOW accidental exceptions, not exceptions you intentionally raised, specifically those occurring after you've already intentionally resolved (either fulfilled or rejected). These are currently "swallowed exceptions" in the truest sense of that concept.

**Comment by bmeurer** [comment]
@getify I guess that should be possible to determine using this hook and the `Error.stack` property, which tells you where the exception originated from (i.e. you can tell from the source location whether it was an explicit `throw` inside of the executor or something else).

**Comment by getify** [comment]
@bmeurer I don't understand. Could you give a quick snippet of code illustrating what you mean?

**Comment by bmeurer** [comment]
@getify Once the swallowed rejection hook is exposed via Node, you should be able to hook into this:

```js
new Promise((res, rej) => {
  // this will raise an error
  return +Symbol.toPrimitive;
});

// Node will fire this hook
function rejectAfterResolved(reason) {
  // Looking at reason.stack you'll see that the error didn't originate from a `throw`
}
```

**Comment by getify** [comment]
> you'll see that the error didn't originate from a `throw`

This is the part I don't understand, specifically. I'm not understanding how the `stack` will give me the information I need. What kind of parsing am I going to have to do here?

**Comment by bmeurer** [comment]
`Error.stack` has column and line number inside, which you can use to find the code that generated the exception. Like if you run this in DevTools:

```js
try { 1+Symbol.toPrimitive; } catch (e) { console.log("" + e.stack); }
```

it'll get back to you with something like

```
VM87:1 TypeError: Cannot convert a Symbol value to a number
    at <anonymous>:1:8
```

The `<anonymous>:1:8` is the information you're interested in.

**Comment by bmeurer** [comment]
@BridgeAR, @benjamingr Do you want to look into adding the hook to Node?

**Comment by benjamingr** [comment]
@bmeurer sure, we need to wait for it to land in Node.js though - it looks like it landed in V8 only a few hours ago IIRC. When Node.js updates to it - I'd love to try adding a check for `kPromiseResolveAfterResolved` and an optional flag - or if @BridgeAR wants to take a stab that's also great :)

**Comment by getify** [comment]
> The `<anonymous>:1:8` is the information you're interested in.

How is an instrumenting library (JS code loaded into a node process) supposed to use that information to determine what the code was? Does such a library have access at runtime to the source code of the program?

I want such a general library to add this hook and start reporting to the console if these exceptions are occurring (but being otherwise swallowed). It needs to be able to work against any arbitrary code, so it won't know anything about the code other than what JS has access to. I'm not sure how, from `error.stack` alone, it would be able to ascertain what that line of code, and column, actually contains?

-------

This also seems like quite a bit of manual work for a user of this hook to have to go to. Is there no way that this hook could just contain a simple flag (or another state) that indicates whether the exception came from an explicit `throw` or not?

**Comment by benjamingr** [comment]
The tl;dr; is you'd use something like [node-stack-trace](https://github.com/felixge/node-stack-trace).

> I want such a general library to add this hook and start reporting to the console if these exceptions are occurring (but being otherwise swallowed)

When we add these hooks to Node.js - you would likely initially get a `--trace-double-resolution` command line flag (or environment variable) that logs a warning with `emitWarning`. 

You will be able to intercept those by adding a `process.on('warning'` handler, filtering by `warning.name` and then access `warning.stack` which should in fact show the line numbers like Benedikt said.

You would then use `node-stack-trace` to `.parse` the error and then have access to `.getFileName()` and `.getLineNumber()` to get the error. 

I think that if we show reasonable adoption for the hook and people do this - it might make sense to expose it natively - at the moment I think we should implement it in Node.js as is as experimental and add capabilities from there.

**Comment by getify** [comment]
> The tl;dr; is you'd use something like node-stack-trace.

I may be misinterpreting those docs, but it seems to suggest that if the stack trace crosses an async boundary, the filename will be like "----------", which doesn't seem like it would serve this use case?

> You would then use node-stack-trace to .parse the error and then have access to .getFileName() and .getLineNumber() to get the error.

OK, I get you can access line number and even filename (at least in some cases -- see above question), but that still doesn't tell me how I would parse the actual JS code that was running. Would I have to use `fs` to read the file in question, parse it for lines, go to a column, and do a regex check to see if `throw` was nearby or not?

I don't get it? This kind of approach sounds deeply manual and intricate.

Why wouldn't it just make sense for v8 to signal with a different ENUM state to indicate whether the exception was explicit (`throw`) or not? v8 must know at that moment. That would be far, far easier, it seems, than trying to later reverse-engineer what manner the exception came from.

**Comment by bmeurer** [comment]
@getify I think what you're proposing is an orthogonal change. If I understand correctly you want to know whether any given `Error` object that was caught at some point eventually originated from a `throw` statement in user code or was generated (implicitly) by the engine?

cc @hashseed

**Comment by getify** [comment]
@bmeurer 

Well... that general of an approach would certainly address my use-case, but... my use-case is much narrower, so it seems overkill.

I'm narrowly only trying to report accidental (not explict) exceptions that occur in a promise constructor after that promise has been resolved (fulfilled or rejected). As far as I know, this is the only case of a truly "accidentally swallowed exception" in JS. I think it was a big oversight in the design of promises to allow such a thing, and I want to surface these errors and potentially eventually be able to advocate for a JS spec change to address them.

An explicit `throw` after promise resolution is also "swallowed" but it wasn't **accidentally** swallowed -- you know what you're doing when you put a `throw` after a resolution. Importantly, those cases are also things that could be caught by a linter or other code-analysis tool.

I am really only trying to address the footgun (which isn't trappable or detectable by any other means) where you are accidentally creating an exception and you have no idea it's happening.

If we solve my use-case with a broader general feature, fine. But I'm not asking for that. It seems to me it would be much easier and more narrow in scope to just solve it with an extra state in this hook.

**Comment by bmeurer** [comment]
@getify We'd need to solve the general problem, since you don't know whether you're executed as *executor* in the `Promise` constructor. I just chatted with @hashseed and we can add such a bit to `Error`. We can talk about this in person next week in Barcelona to figure out the details.

**Comment by mike-kaufman** [comment]
I briefly skimmed discussion here, so apologies if I missed this, but per spec, a promise's state must be pending when you call fulfill/reject (see [here](https://tc39.github.io/ecma262/#sec-fulfillpromise) and [here](https://tc39.github.io/ecma262/#sec-rejectpromise)) - first step of the spec'd algorithms make this assertion.  Rather than adding additional callbacks, isn't it simpler here for engine to throw if caller tries to reject/resolve a promise that isn't in pending state?  Or are there breaking change concerns with this approach?

**Comment by getify** [comment]
> isn't it simpler here for engine to throw if caller tries to reject/resolve a promise that isn't in pending state?

What would the engine or environment do with this exception? Since it's inside of a promise constructor/executor, all exceptions are supposed to turn into promise rejections. But, promises aren't supposed to change states after fulfillment, so the engine couldn't just change the promise's state to rejected for that exception. So there needs to be a third "channel" for those to be reported, which is the purpose of this hook.

The fundamental problem here is that the promise spec either didn't contemplate, or did contemplate and chose not to worry about, what happens if you try to fulfill after fulfillment. That "extra state" is just ignored. An extra success/resolve is not such a big deal if it's swallowed -- at least, one could understand why it's swallowed. An extra `reject(..)` will also be swallowed, which is unfortunate but similarly justifiable.

BUT, promises currently will swallow an actual exception (either from `throw ..` or an accidental exception) in an entirely uncatchable way. That's what we're trying to solve because at least some of us find that intolerable.

**Comment by devsnek** [comment]
V8 now has a way to track resolve after resolve and reject after resolve (no fulfill after resolve, which just has to do with the structure of resolve functions in the spec) which we are landing onto node in https://github.com/nodejs/node/pull/21838

now that the engine(s) provide a way to hook this, perhaps the discussion can shift into how it should be exposed to users? (personally i think node should just re-emit the V8 stuff and let userland create the behaviour on its own but that's just me)

**Comment by mike-kaufman** [comment]
> what would the engine or environment do with this exception?

Exception would just be a normal exception - if there's an exception handler on the stack, then it would get invoked.  If not, then host's unhandled exception machinery would kick in.   I don't believe there's anything in spec that says a promise ctor should execute async (plz lmk if I'm wrong).  Perhaps there are more complex cases where this doesn't work? 

Looking through spec some more, I think the problem is the spec states calling reject/resolve on an already resolved promise just returns undefined (see step 4 [here](https://tc39.github.io/ecma262/#sec-promise-reject-functions) and [here](https://tc39.github.io/ecma262/#sec-promise-resolve-functions)).

This seems weird.  I'd like to know if existing language is here for a reason.

The thing that bothers me with the callback approach here is it is a notification mechanism for misuse of an API; exceptions fit this use case nicely. That is, behavior should be no different than, say calling `someArray.filter('a'); // 'a' is not a function and filter will throw`,

**Comment by devsnek** [comment]
@mike-kaufman 

> I'd like to know if existing language is here for a reason.

basically that is how promises only resolving once is enforced. if you keep moving on to [FullfillPromise](https://tc39.github.io/ecma262/#sec-fulfillpromise) and [RejectPromise](https://tc39.github.io/ecma262/#sec-rejectpromise) you'll notice they start with `Assert: The value of promise.[[PromiseState]] is "pending".`.

it is expected that (like V8 has done) implementations can bring their own debug tooling to the table.

as a sidenote, perhaps [HostPromiseRejectionTracker](https://tc39.github.io/ecma262/#sec-host-promise-rejection-tracker) can be reworked to provide information about resolve/reject after resolve but i don't think its needed since like i said engines can already bring their own debug

**Comment by mike-kaufman** [comment]
> basically that is how promises only resolving once is enforced. 

Sure, but throwing here (vs returning undefined) could also enforce that promises are only resolved once, right?  IMO, the callback approach feels like an indirect and inconsistent way to notify a user of API misuse.  Moreover, if exceptions raised in resolve/reject "won't work" because propagation is busted, then that holds for exceptions on any other API.

It's not my intention to derail progress here, I'm just trying to understand why there isn't a solution more consistent with existing language constructs.

> as a sidenote, perhaps HostPromiseRejectionTracker can be reworked to provide information about resolve/reject after resolve but i don't think its needed since like i said engines can already bring their own debug

Don't know that I agree w/ that.  This is effectively a notification mechanism about a misused API, which seems spec-worthy. Or am I missing something?

**Comment by benjamingr** [comment]
> Don't know that I agree w/ that. This is effectively a notification mechanism about a misused API, which seems spec-worthy. Or am I missing something?

Node.js is at liberty to warn users against mistakes here. This is very consistent with language constructs and desirable. This change can save people a lot of time debugging in the future.

**Comment by getify** [comment]
> I don't believe there's anything in spec that says a promise ctor should execute async (plz lmk if I'm wrong). 

Nope, its definitely synchronous.

And that's the problem, because we cannot throw a synchronous exception from the promise constructor/executor -- the invariant behavior guaranteed by promises is that the constructor (at least if given a valid executor function) always runs successfully, and any exceptions inside it are turned into promise rejections. It would definitely break a lot of code if synchronous exceptions starting emitting in places where people were *promised* they wouldn't.

So the only other option would be to force it as an "unhandled exception". AFAIK, there's no other place where we have synchronous calls that can create, in essence, asynchronous exceptions... so that option seems weird to me.

> it is a notification mechanism for misuse of an API

Not necessarily misuse. The accidental exception case (the one I care most about) is not misuse of the API at all. It's just an unintended mistake that unfortunately happens in a dead zone.

**Comment by benjamingr** [comment]
It's worth mentioning that adding this hook (which @TimothyGu is PRing the V8 changes for in core right now) enables people to experiment with the behaviour. 

This would let us prototype and see how users use it - maybe a warning is also a good first step.

**Comment by mike-kaufman** [comment]
> Node.js is at liberty to warn users against mistakes here. This is very consistent with language constructs and desirable. This change can save people a lot of time debugging in the future.

So, I'm not claiming Node.js can't do anything here and I'm not claiming that this isn't useful to end users.  And I'm not suggesting that the approach suggested in this thread shouldn't be done - if that works for everyone then go for it.

To be clear, what I am claiming is: 
  - calling fulfill/reject on a resolved promise is an error condition (despite current spec language)
  - the error condition is immediate at the call site (vs a "potential error", like unhandled rejections). 
  - the consistent approach for such "immediate" errors is raising an exception.
  - the solution proposed here seems like a work-around to a spec bug.  Namely, that exceptions occurring in promise ctor/executor are potentially swallowed.
  - The solution here doesn't necessarily accrue to the browser in a consistent way.
 
Now, I've been wrong before, and I'm happy to be proven wrong on any of the above, so please let me know if I'm missing something. :)

It feels like the most impactful approach here is to get the spec fixed.  In particular, 
  - https://tc39.github.io/ecma262/#sec-promise-executor, step 10 
    - change to account for abrupt completions when promise is not in pending state.
  - https://tc39.github.io/ecma262/#sec-promise-reject-functions, step 4
    - change to throw instead of returning undefined
  - https://tc39.github.io/ecma262/#sec-promise-resolve-functions, step 4
    - change to throw instead of returning undefined

I'm sure folks here are more familiar w/ the spec than I am, so please let me know if the above is overly simplistic.

**Comment by benjamingr** [comment]
> the solution proposed here seems like a work-around to a spec bug. Namely, that exceptions occurring in promise ctor/executor are potentially swallowed.

We don't have consensus that it's a spec bug and there might be code in the wild that uses this behaviour. Note the discussion on the google doc that Benedikt linked to and the discussion there. 

We can raise spec changes or bring the issue for discussion with TC39 if you'd like but I'd rather not ask Daniel for more spec discussions before we are convinced internally it makes sense to change the spec itself.

**Comment by mike-kaufman** [comment]
> We don't have consensus that it's a spec bug

:) .... swallowed exceptions sure sounds like a spec bug to me. 

Breaking changes are absolutely a concern, but I don't know that's sufficient to not do this the "right way".

**Comment by mike-kaufman** [comment]
@getify 

> And that's the problem, because we cannot throw a synchronous exception from the promise constructor/executor

Why not? :)

As an aside, trying to make a distinction between syntactic throws vs vm-level throws feels wrong.  What are you trying to achieve here?

**Comment by devsnek** [comment]
@mike-kaufman the design goal is that once a promise goes the only way to interact with it is via the one interface (.then, .catch, .finally) to streamline interaction. `try { new Promise(...).catch(...) } catch(e) { ... }` isn't a very nice pattern.

as an aside, if you want to play with promise behaviour i maintain a spec compliant reference implementation written in javascript which is pretty easy to mess with: https://github.com/devsnek/promise-reference

**Comment by mike-kaufman** [comment]
> the design goal is that once a promise goes the only way to interact with it is via the one interface (.then, .catch, .finally) to streamline interaction.

Sure, so then https://tc39.github.io/ecma262/#sec-promise-executor  wording needs to account for abrupt completions in a way that will cause the promise to be rejected, no?

**Comment by devsnek** [comment]
@mike-kaufman step 10 already does that. the promise just can't change its state. that's why i said above that this stuff is best left implementation-specific debugging.

**Comment by mike-kaufman** [comment]
> step 10 already does that

My interpretation here (perhaps wrong?) is the *intent* with step 10 is to reject the promise.  However, if the promise is already resolved, then step 10 effectively no-ops because of  step 4  in the reject/resolve abstract operations.   

I'm having a hard time seeing how spec language that allows for lost exceptions is any way desirable or intentional.  I get the idea that exceptions shouldn't propagate out of the executor, but they shouldn't be lost.

**Comment by devsnek** [comment]
> but they shouldn't be lost.

right. that's what debugging is for. i assume your production code's intention isn't to double-resolve.

**Comment by getify** [comment]
> > we cannot throw a synchronous exception from the promise constructor/executor

> Why not? :)

As explained thereafter, because that breaks the invariant behavior that people know and expect from promises. There's a 100% chance that this would be an intolerable backwards-breakage to just start throwing synchronously at the site of `new Promise(..)`.

> ...trying to make a distinction between syntactic throws vs vm-level throws feels wrong.  What are you trying to achieve here?

I consider a `reject(..)` call and `throw ..` as both being **explicit** attempts to reject a promise. If you intentionally try to reject a promise that's already been fulfilled, that's you misusing the API. We can debate now, academically, what that outcome should be, but the spec already long ago decided that this case was, "we just silently swallow your intentional misuse". I am fairly certain though that we cannot change it without intolerable backwards breakage.

However, an accidental exception in that same position, I argue, is not intentional misuse. It's an accident like a typo, like typing `consloe.log(..)` (instead of `console.log(..)`). I'm arguing that accidental exceptions being swallowed are devious enough that they argue more strongly for special handling than the `reject` / `throw` do.

For example, I can imagine a developer (me) wanting a secondary `reject` / `throw` to be reported as a "Misused the API" exception. But at the same time, they may want the secondary accidental exceptions to be handled differently. I'd want accidental exceptions to be able to override the promise state and force it to be rejected with the exception, no matter what its previous fulfillment state was. I feel that way because accidental exceptions should never, ever be hidden and I want JS to be very vocal and forceful in letting me know about them. By contrast, misuse of the API is a lesser offense, and I want to know but I don't need the engine to stop everything to do so.

These are all judgement calls I'm personally making. I realize others may not agree. But the whole point of this thread and feature request is that I want the ability to subjectively make those distinctions for my own coding, even if I'm the only one.

FYI: among many other experiments, overriding the promise state in the case of a secondary exception (but not `reject(..)`) is implemented in [this experimental `BetterPromise(..)` subclass](https://gist.github.com/getify/1173cac45d15fc4ff0a880f32fd598ab), for the purposes of exploring whether those semantics are useful in real-world code. Sadly, I cannot (yet) distinguish between `throw` and accidental exceptions like I would prefer to do.

**Comment by devsnek** [comment]
there are also a lot of timing considerations that went into the design of promises.

consider the following:

(`process.nextTick` runs before `PromiseJobs` and the `setTimeout` runs after)

```js
new Promise((resolve, reject) => {
  setTimeout(resolve, 2);
  process.nextTick(reject);
  resolve();
});
```

or a less contrived example:

```js
Promise.race([ // <- racing depends on resolution calls!
               // this is impossible if calling resolve twice
               // does something besides returning undefined.
  new Promise((r) => r()),
  new Promise((r) => process.nextTick(r));
  new Promise((r) => setTimeout(r, 2));
]);
```

**Comment by getify** [comment]
> I'm having a hard time seeing how spec language that allows for lost exceptions is any way desirable or intentional. I get the idea that exceptions shouldn't propagate out of the executor, but they shouldn't be lost.

I have personally brought this issue up years ago, several times, with multiple members of TC39, prior to the finalization of ES6. It landed this way anyway. So I'm pretty certain that this wasn't an oversight on their part, but intentional.

Perhaps asking @allenwb @domenic and others if they have any old notes or recollections from those TC39 discussions?

**Comment by allenwb** [comment]
@getify I don't have time to digest this whole thread right now. But here is what immediately comes to mind about such "lost exceptions".  My recollection is that in the ES6 specification (things might have subsequently changed) an uncaught exception bubbles as an completion value all the way up to whatever implementation code initiated the job that is running when the exception is thrown.  For exceptions that occur in promised related jobs the exceptions would propagate to the part of the implementation that schedules and initiates those jobs. It is left up to implementations to decide how they want to deal with such unhandled exception.  Simply ignoring them would be the last thing I would expect an implementation to do.

**Comment by devsnek** [comment]
@allenwb we're specifically talking about errors inside the promise constructor and `PromiseJobs` queue executions, where exceptions after the promise is resolved don't change the state of the promise or provide a way to be detected at all. sorry for the loss of context.

**Comment by allenwb** [comment]
I know, but unhandled exceptions in jobs dispatched off of the promise job queue work the same way. The propagate to the implementation code that dispatched the job so it is up to the implementation as to what to do. I would expect an implementation to at least log the exception to its console.

It certainly would be possible for the ES spec. to specify some more specific handling, but what should it be?  A lot of possibilities come mind but this seems like a situation where implementation experimentation needs to be done before something is standardized.

Separately, there was quite a bit of discussion about what should be done with errors that are detected in the constructor.  Some people (including me) thought that API level errors should have thrown an exception rather than rejecting the new promise.  But most people (or at least the loudest of those that cared) wanted wanted a rejection so that's what ended up going into the specification.

**Comment by mike-kaufman** [comment]
@allenwb - 

Thanks for chiming in.  To clarify, there are two issues here that would be good to understand the back-story on:

1.  In the following code, the raised exception is swallowed, by spec - no exception propagates out of the constructor & the resulting promise is not rejected.

```javascript
  let p = new Promise((resolve, reject) => {
      resolve();
      throw new Error();
  }
```

2.  calling resolve()/reject() on a resolved promise returns undefined (instead of throwing)

**Comment by mike-kaufman** [comment]
@devsnek - thanks for examples.  I'm assuming that spec's race text could be updated to account for abnormal completions on resolve.  Of course, such a change would break 3rd party implementations, but I'm not clear on the impact.   RE your contrived example, how prevalent are such patterns and are they bugs or intentional?

**Comment by mike-kaufman** [comment]
@getify  - thanks for clarifying your intent here.  I get it, but as you noted above, you have to go through some unnatural contortions to understand the distinctions you want, which feels...  error prone & hacky.  Altenative is VM support for specific bits to indicate source of exception, which means we've now got user code taking VM-specific deps, which I'd prefer to avoid.

**Comment by devsnek** [comment]
@mike-kaufman well without the current behaviour Promise.race is broken.

aside from that, i don't think it would be good design for there to be a difference between

```js
new Promise((r) => {
  setTimeout(r, 10, 2);
  r(1);
}).then(???);

new Promise((r) => {
  process.nextTick(r, 2);
  r(1);
}).then(???);

new Promise((r) => {
  r(1);
  r(2);
}).then(???);

new Promise((r) => {
  r(1);
  throw new Error('2');
});

Promise.race([
  new Promise((r) => r(1)), // could be two http
  new Promise((r) => r(2)), // requests, who knows...
]).then(???);
```

if you can describe a regular pattern for those that isn't the current behaviour then we've solved the problem

if there is no other option, then it has to be left to debug tooling

**Comment by allenwb** [comment]
@mike-kaufman 
I see, what you are really commenting upon is that the built-in [reject function](https://tc39.github.io/ecma262/#sec-promise-reject-functions) that is passed to an executor silently does nothing if its associated promise is already resolved.  This happens regardless of whether it is called from within the promise constructor or sometime later after the constructor returns.  The exception in the example is dropped because the standard promise constructor turns exceptions propagated from the executor into calls to the reject function and in this example the promise has already been resolved.

This is a consequence of the general Promise API design decision that exceptions should all be turned into rejections. An alternative design might have had the standard constructor propagate such unhandled executor exceptions instead of doing a reject. Or perhaps only doing the reject if the promise was not already resolved.

I don't recall whether this specific cases was discussed within TC39, if it was the discussion would probably have been captured in the meeting notes from circa 2013-2015.  You might also be able to track the evolution of the specification for the reject function by looking at the intermediate spec drafts captured by the [Web Archive](https://web.archive.org/web/20160305002637/http://wiki.ecmascript.org/doku.php?id=harmony:specification_drafts).

Note that there is nothing in the specification of the reject function that prevents an implementation from doing some additional, non-intrusive action when a reject function is called on an already resolved promise. For example, an implementation certainly is allowed to log such an occurrence.

@getify It seems to me that a better promises subclass with some cleaver wrapping of the user provided executor and the implementation generated resolve/reject functions could change this behavior for user code calls to resolve/reject.  But that would not catch internally specified calls to the the built-in resolve/reject functions that are attached  to instances of such subclasses.  This seems like a bug in the subclassability design of Promise. I haven't search to see if there are any other specified calls to a built-in reject that would have this same problem.  If the only one is in the Promise constructor you might be able to work around it by making sure that your executor wrapper catches all exceptions.  It could then use some sort of side channel to communicate the exception back to the better promise constructor.

**Comment by benjamingr** [comment]
@larspars ping regarding your medium article - we acknowledge the problem you raised there of resolving after rejecting and are providing instrumentation for it :) Since you care about this do feel free to get involved.

**Comment by larspars** [comment]
Thanks for the ping @benjamingr. 

I'm very much in favor of issuing a warning in these cases of double reject/resolve. 

In my post on [error handling in Promises](https://hackernoon.com/n²-ways-to-shoot-yourself-in-the-foot-error-handling-with-async-await-and-promises-87f30cda1681), I'm hinting at how I would have ideally liked this to be designed, but I guess I don't explain it in too much detail.

I'm aware the following it's a fairly drastic proposal, but I'm not yet convinced it is crazy-talk. This will be a bit long, but I think it's relevant to the discussion.

My take on this is that the complexities we're talking about here are not going to be top-of-mind for 99% of programmers using Promises. They'll be unaware of the case that you can do both reject and resolve on the same Promise. They'll be unaware of when a Promise rejection can be considered unhandled. Etc. So most programmers will be blind to a bunch of problems that may happen.

The solution I propose is to throw away resolve/reject entirely, in favor of return/throw. Then you get language level support for there being only one of these happening for a given Promise. In other words, we stop doing this:

    return new Promise((resolve, reject) => {
        if (thing.isAThing) { resolve(thing); }
        if (thing.hasErrors) { reject(thing.accumulatedErrors); }
    })
   
And only do it like this:

    return new Promise(() => {
        if (thing.isAThing) { return thing; }
        if (thing.hasErrors) { throw new Error(thing.accumulatedErrors); }
    })
    
This removes a whole class of problems. But, it opens a new question: How would you implement rejects and resolves from async functions? I.e. how would you implement this:

    return new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve("ok");
        }, 100);
    })
    
Here, I propose that this is done with a new variant of setTimeout, i.e:

    return new Promise(() => {
        return await setTimeout2(() => {
            return "ok";
        }, 100);
    })
    
Here setTimeout2(fn) has the semantics that:

1) It returns whatever fn returns.
2) It rethrows whatever fn throws.
3) It's awaitable

I initially thought this would need runtime support, but I think it's doable in user land. Then you don't need reject/resolve any more. There is a small number of use cases that go away. For example, you can't do this:

    return new Promise((resolve, reject) => {
        button.onclick = () => resolve("button was clicked");
    })
    
I think that's fine. There's a small loss of generality, in favor of a serious reduction in complexity. I think that's exactly what one should want from a design. It's exactly why array.forEach() is nicer than a straight for-loop, for example.

It's also clear that such a design would work, because it's exactly the design of the [.NET Task Parallel Library](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/task-parallel-library-tpl).

The reduced complexity from this would be pretty massive. This entire discussion thread would be uneeded, and it seems like almost everything that is complex about Promises goes away.

It's a fairly radical proposal. If this was five years ago, and Promises was not implemented in browsers, I would argue strongly in its favor. Today, I still think it's a better design, but it's possible the cost of switching to it may be too high.

Actually removing resolve/reject seems like a no-go at this point, but perhaps one could move forward by downplaying resolve/reject in favor of using return/throw along with a new API for setTimeout().

Assuming everyone buys into this being a better design (big assumption), there are still some questions:
- What conflicts, if any, exist with other parts of the current Promises API?
- What's a good way to discourage use of reject/resolve? Issuing warnings when they are called for the first time?
- Given that the reject/resolve design already exists, is it even worth it to push this other way of doing things?

**Comment by benjamingr** [comment]
@larspars I think you're barging into an wide open door :)

The tl;dr; is that we agree, we want to provide APIs so that people never _have_ to write `new Promise` and we want to provide hooks to improve debugging support for the promise constructor anyway just in case (like the case you raised in your article).

-----

Ideally I want people to never have to use the promise constructor at all - that's why we have `util.promisify` and promise APIs. Ideally, people will use `async function`s and not "raw" promises in their code when possible and engines will provide better debugging tooling for those like async stack traces.

```js
return new Promise(() => {
    if (thing.isAThing) { return thing; }
    if (thing.hasErrors) { throw new Error(thing.accumulatedErrors); }
})
```

Can literally be `return` and `throw` inside an `async function`.

```js
let delay = util.promisify(setTimeout); // Added in Node 8
async function doFoo() {
   await delay(100);
   // can `throw` or `return` here.
}
doFoo();
```

I couldn't agree more with:

> My take on this is that the complexities we're talking about here are not going to be top-of-mind for 99% of programmers using Promises. They'll be unaware of the case that you can do both reject and resolve on the same Promise. They'll be unaware of when a Promise rejection can be considered unhandled. Etc. So most programmers will be blind to a bunch of problems that may happen.

(We do intend to do a survey to validate it)

Ideally - just like in .NET where people don't use `TaskCompletionSource` but use APIs that already return promises. We already have this experimentally for `fs` and `dns` is coming next. We also have async iterator (for... await) support for streams. Hopefully in a year or two we'll have such APIs for all the relevant APIs in Node.

It _is_ somewhat radical to provide promise APIs for everything - but given that's what people use and that's what the language does - we're still going to do it.

**Comment by larspars** [comment]
@benjamingr Aha, sorry if I was barging in and stating what everyone was already thinking, haha. I haven't followed the discussions around this, but what you describe looks really nice, and looks like it would solve the problems I was talking about.

**Comment by benjamingr** [comment]
@larspars if you have any other ideas please do let me know, personally I posted [a canonical Q&A](https://stackoverflow.com/questions/23803743/what-is-the-explicit-promise-construction-antipattern-and-how-do-i-avoid-it) about it - proposed this [hook](https://github.com/nodejs/promise-use-cases/blob/d58611e379a599d655c5f917f543e83c7821e75a/use-cases/user-expectations/6/expectations-6.md) in the Node collaborator summit, proposed [.promisify](https://github.com/nodejs/CTC/issues/12) (and pushed for it and bluebird before it was native) and am trying to promote the new promise APIs where I can.

I don't think that's enough (or that we're moving quickly enough) which is why I think a lot of people are running into the same issues you are (and aren't writing blog posts about them or are as willing to interact).

I personally don't believe we're in a very good place with promises and how we work with them (yet) - so any suggestions for hooks, debugging ideas etc. are welcome. Benedikt here is in V8 and while this is a node repository I would also happily connect you to relevant individuals in different browsers or the spec if you can prototype with hooks and provide feedback to implementors with CatchJS (or as an individual).

**Comment by larspars** [comment]
@benjamingr Sweet, I'll look over your links and let you know if I think I have anything semi-intelligent to contribute :)

**Comment by benjamingr** [comment]
@larspars sure, just don't assume we've thought of most of this stuff or have done so correctly - if you have ideas please do bring them up by opening an issue here or in nodejs/node
